File: frontend/src/api/workforce.ts
import { apiClient } from './client'

export interface EmployeeRow {
  name: string
  employee_name?: string
  department?: string
  designation?: string
  company?: string
  status?: string
  doctype: 'Employee'
  label?: string
}

export interface AttendanceRow {
  name: string
  employee?: string
  employee_name?: string
  attendance_date: string
  department?: string
  status?: string
  company?: string
  doctype: 'Attendance'
  label?: string
}

export interface LeaveApplicationRow {
  name: string
  employee?: string
  employee_name?: string
  department?: string
  leave_type?: string
  from_date: string
  to_date?: string
  status?: string
  doctype: 'Leave Application'
  label?: string
}

export interface ShiftTypeRow {
  name: string
  start_time?: string
  end_time?: string
  enable_auto_attendance?: number
  disabled?: number
  doctype: 'Shift Type'
  label?: string
}

export interface ShiftAssignmentRow {
  name: string
  employee?: string
  employee_name?: string
  shift_type?: string
  start_date: string
  end_date?: string
  status?: string
  reconciliation_status?: 'Open' | 'Closed' | 'Verified'
  total_expected_sales?: number
  closed_at?: string
  doctype: 'Shift Assignment'
  label?: string
}

export interface CompanyRow {
  name: string
  company_name?: string
  doctype: 'Company'
  label?: string
}

interface ListWorkforceResponse<T> {
  doctype: string
  count: number
  records: T[]
}

const listWorkforceRecords = async <T>(
  doctype: 'Employee' | 'Attendance' | 'Leave Application' | 'Shift Type' | 'Shift Assignment' | 'Company',
  params: Record<string, any> = {},
) => {
  const response = await apiClient.get('/api/method/national_oil.api.workforce.list_workforce_records', {
    params: {
      doctype,
      limit_page_length: 50,
      ...params,
    },
  })

  return (response.data.message as ListWorkforceResponse<T>).records
}

export const workforceApi = {
  async listEmployees() {
    return await listWorkforceRecords<EmployeeRow>('Employee')
  },

  async listAttendance(status?: string) {
    return await listWorkforceRecords<AttendanceRow>('Attendance', {
      ...(status ? { status } : {}),
    })
  },

  async listLeaveApplications(status?: string) {
    return await listWorkforceRecords<LeaveApplicationRow>('Leave Application', {
      ...(status ? { status } : {}),
    })
  },

  async listShiftTypes() {
    return await listWorkforceRecords<ShiftTypeRow>('Shift Type')
  },

  async listShiftAssignments(status?: string) {
    return await listWorkforceRecords<ShiftAssignmentRow>('Shift Assignment', {
      ...(status ? { status } : {}),
      limit_page_length: 100,
    })
  },

  async listCompanies() {
    return await listWorkforceRecords<CompanyRow>('Company')
  },
}

File: national_oil/api/workforce.py
import frappe


ALLOWED_WORKFORCE_DOCTYPES = {
	"Employee": {
		"label_field_candidates": ["employee_name", "name"],
		"fields": ["name", "employee_name", "department", "designation", "company", "status"],
		"search_fields": ["name", "employee_name", "department", "designation", "company"],
	},
	"Driver": {
		"label_field_candidates": ["driver_name", "full_name", "name"],
		"fields": [
			"name",
			"driver_name",
			"full_name",
			"employee",
			"transporter",
			"supplier",
			"phone",
			"cell_number",
			"license_number",
			"status",
		],
		"search_fields": ["name", "driver_name", "full_name", "phone", "cell_number", "license_number"],
	},
	"Attendance": {
		"label_field_candidates": ["employee_name", "employee", "name"],
		"fields": ["name", "employee", "employee_name", "attendance_date", "department", "status", "company"],
		"search_fields": ["name", "employee", "employee_name", "department", "status"],
	},
	"Leave Application": {
		"label_field_candidates": ["employee_name", "employee", "name"],
		"fields": [
			"name",
			"employee",
			"employee_name",
			"department",
			"leave_type",
			"from_date",
			"to_date",
			"status",
		],
		"search_fields": ["name", "employee", "employee_name", "department", "leave_type", "status"],
	},
	"Shift Type": {
		"label_field_candidates": ["name"],
		"fields": ["name", "start_time", "end_time", "enable_auto_attendance", "disabled"],
		"search_fields": ["name"],
	},
	"Shift Assignment": {
		"label_field_candidates": ["employee_name", "employee", "name"],
		"fields": ["name", "employee", "employee_name", "shift_type", "start_date", "end_date", "status"],
		"search_fields": ["name", "employee", "employee_name", "shift_type", "status"],
	},
}


def _get_config(doctype: str) -> dict:
	base_config = ALLOWED_WORKFORCE_DOCTYPES.get(doctype)
	if not base_config:
		frappe.throw(frappe._("DocType {0} is not allowed for workforce lookup").format(doctype))
	if not frappe.db.exists("DocType", doctype):
		frappe.throw(frappe._("DocType {0} is not installed on this site").format(doctype))
	if not frappe.has_permission(doctype, "read"):
		frappe.throw(frappe._("Not permitted to read {0}").format(doctype), frappe.PermissionError)

	meta = frappe.get_meta(doctype)
	available_fields = {"name"} | {field.fieldname for field in meta.fields}
	label_field = next(
		(fieldname for fieldname in base_config["label_field_candidates"] if fieldname in available_fields),
		"name",
	)

	return {
		"label_field": label_field,
		"fields": [fieldname for fieldname in base_config["fields"] if fieldname in available_fields],
		"search_fields": [
			fieldname for fieldname in base_config["search_fields"] if fieldname in available_fields
		],
	}


def _build_or_filters(search_fields: list[str], txt: str) -> list[list[str]]:
	if not txt:
		return []
	return [[field, "like", f"%{txt}%"] for field in search_fields]


@frappe.whitelist()
def list_workforce_records(
	doctype,
	txt=None,
	employee=None,
	department=None,
	status=None,
	date_from=None,
	date_to=None,
	limit_page_length=20,
):
	"""Return standardized workforce/logistics records from ERPNext and HRMS."""
	doctype = (doctype or "").strip()
	txt = (txt or "").strip()
	limit_page_length = min(max(int(limit_page_length or 20), 1), 100)

	config = _get_config(doctype)
	fields = list(dict.fromkeys(config["fields"] + [config["label_field"]]))
	filters = {}
	if employee:
		filters["employee"] = employee
	if department:
		filters["department"] = department
	if status:
		filters["status"] = status
	date_field = None
	if doctype == "Attendance":
		date_field = "attendance_date"
	elif doctype == "Leave Application":
		date_field = "from_date"

	if date_field:
		if date_from and date_to:
			filters[date_field] = ["between", [date_from, date_to]]
		elif date_from:
			filters[date_field] = [">=", date_from]
		elif date_to:
			filters[date_field] = ["<=", date_to]

	records = frappe.get_all(
		doctype,
		fields=fields,
		filters=filters,
		or_filters=_build_or_filters(config["search_fields"], txt),
		limit_page_length=limit_page_length,
		order_by="modified desc",
	)

	for record in records:
		record["label"] = record.get(config["label_field"]) or record.get("name")
		record["doctype"] = doctype

	return {
		"doctype": doctype,
		"count": len(records),
		"records": records,
	}


@frappe.whitelist()
def get_workforce_lookup_bundle():
	"""Return installed canonical workforce/logistics doctypes for the frontend."""
	bundle = []
	for doctype, config in ALLOWED_WORKFORCE_DOCTYPES.items():
		if frappe.db.exists("DocType", doctype) and frappe.has_permission(doctype, "read"):
			resolved = _get_config(doctype)
			bundle.append(
				{
					"doctype": doctype,
					"label_field": resolved["label_field"],
					"fields": resolved["fields"],
				}
			)
	return bundle
